rs/replicated_state/src/canister_state/queues/queue.rs

```diff
@@ -109,17 +109,22 @@ pub(crate) struct CanisterQueue {
     capacity: usize,
 
     /// Number of enqueued request references.
+    ///
+    /// Invariants:
+    ///  * `request_slots == queue.iter().filter(|item| !item.is_response()).count()`
+    ///  * `request_slots <= capacity`
     request_slots: usize,
 
     /// Number of slots used by response references or reserved for expected
     /// responses.
+    ///
+    /// Invariants:
+    ///  * `response_slots >= queue.iter().filter(|item| item.is_response()).count()`
```

Comment:
Should we then called these `reserved_response_slots`?

Reply:
Not really. We could have had two separate counters, one for actually enqueued (including potentially stale) responses; and one for reserved slots. But it would have taken up more space and would have e.g. required decrementing the number of reserved slots and incrementing the number of enqueued responses whenever we did enqueue a response.

I.e. this is a count of how many response slots are in use. If it's equal to the capacity, we can make no more reservations (backpressure). If it plus the number of request slots is equal to the queue size, it means these are all actual enqueued responses and we have zero reservations available.